Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs

```diff
@@ -92,7 +121,105 @@ public void Dispose()
+                throw new QldbDriverException(ExceptionMessages.DriverClosed);
+            }
+
+            bool replaceDeadSession = false;
+            int retryAttempt = 0;
+            while (true)
+            {
+                AsyncQldbSession session = null;
+                try
+                {
+                    if (replaceDeadSession)
+                    {
+                        session = await this.StartNewSession();
+                    }
+                    else
+                    {
+                        session = await this.GetSession();
+                    }
+
+                    T returnedValue = await session.Execute(func);
+                    this.ReleaseSession(session);
+                    return returnedValue;
+                }
+                catch (RetriableException re)
+                {
+                    // If initial session is invalid, always retry once with a new session.
+                    if (re.InnerException is InvalidSessionException && retryAttempt == 0)
+                    {
+                        this.logger.LogDebug("Initial session received from pool invalid. Retrying...");
+                        replaceDeadSession = true;
+                        retryAttempt++;
+                        continue;
+                    }
+
+                    // Normal retry logic.
+                    if (retryAttempt++ >= retryPolicy.MaxRetries)
+                    {
+                        if (re.IsSessionAlive)
+                        {
+                            this.ReleaseSession(session);
+                        }
+                        else
+                        {
+                            this.poolPermits.Release();
+                        }
+
+                        throw re.InnerException;
+                    }
+
+                    this.logger.LogInformation("A recoverable error has occurred. Attempting retry #{}.", retryAttempt);
+                    this.logger.LogDebug(
+                        "Errored Transaction ID: {}. Error cause: {}",
+                        re.TransactionId,
+                        re.InnerException.ToString());
+                    replaceDeadSession = !re.IsSessionAlive;
+                    if (replaceDeadSession)
+                    {
+                        this.logger.LogDebug("Replacing invalid session...");
+                    }
+                    else
+                    {
+                        this.logger.LogDebug("Retrying with a different session...");
+                        this.ReleaseSession(session);
+                    }
+
+                    try
+                    {
+                        var backoffDelay = retryPolicy.BackoffStrategy.CalculateDelay(
+                            new RetryPolicyContext(retryAttempt, re.InnerException));
+                        await Task.Delay(backoffDelay, cancellationToken);
+                    }
+                    catch (Exception e)
```

Comment:
Will this realistically throw Exception, or will it throw a specific exception type?

Reply:
This is a safeguard against people throwing exceptions in their backoffDelay for whatever reason.